core/src/main/scala/cats/Traverse.scala

```diff
@@ -135,12 +135,24 @@ import simulacrum.typeclass
   override def map[A, B](fa: F[A])(f: A => B): F[B] =
     traverse[Id, A, B](fa)(f)
 
+  /**
+   * Akin to [[map]], but allows to keep track of a state value
+   * when calling the function.
+   */
+  def mapAccumulate[A, S, B](fa: F[A])(init: S)(f: (A, S) => (B, S)): (F[B], S) =
+    traverse(fa) { a =>
+      State.get[S].flatMap { s1 =>
+        val (b, s2) = f(a, s1)
+        State.set(s2).map(_ => b)
+      }
+    }.run(init).value.swap
```

Comment:
I wonder if this can be simplified:
```suggestion
    traverse(fa) { a =>
      State { (s: S) => f(a, s).swap }
    }.run(init).value.swap
```